enhancements/update/dont-require-registry-during-reboot-and-upgrade.md

```diff
@@ -0,0 +1,198 @@
+replaces: []
+superseded-by: []
+---
+
+# Don't require registry during reboot and upgrade
+
+## Summary
+
+Ensure that clusters don't require a registry server to reboot existing nodes or
+upgrade when all the required images have already been pulled.
+
+## Motivation
+
+Currently during reboots of existing nodes and upgrades clusters may need to
+contact the image registry servers, even if the images have already been pulled.
+This complicates things for clusters that are completely disconnected or that
+have an slow or unreliable connection to the image registry servers.
+
+### User Stories
+
+#### Reboot without registry
+
+As the administrator of a cluster that has all the required images already
+pulled in all the nodes, I want to be able to reboot nodes it without requiring
+access to a registry server.
+
+### Upgrade without registry
+
+As the administrator of a cluster that has all the required images already
+pulled in all the nodes, I want to be able to upgrade it without requiring
+access to a registry server.
+
+### Goals
+
+Ensure that clusters don't require a registry server to reboot nodes or upgrade
+when all the required images have already been pulled.
+
+### Non-Goals
+
+It is not a goal of this enhancement to add a mechanism to ensure that
+required images are available. That is the subject of the [pin and pre-load
+images](https://github.com/openshift/enhancements/pull/1481) enhancement.
+
+It is not a goal of this enhancement to eliminate the requirement of a registry
+server for other workloads. We are just eliminating the requirement that it be
+available during reboots of existing nodes and upgrades.
+
+## Proposal
+
+### Workflow Description
+
+1. The administrator of a cluster reboots a node or performs an upgrade.
+
+1. All the components of the cluster ensure that if the required images have
+been already pulled they will not try to contact the registry server.
+
+### API Extensions
+
+None.
+
+### Implementation Details/Notes/Constraints
+
+#### Don't use the `Always` pull policy
+
+Some OCP components currently use the `Always` image pull policy during
+upgrades. As a result, the kubelet and CRI-O will try to contact the registry
+server, even if the image is already available in the local storage of the
+cluster. This blocks upgrades and should be avoided.
+
+Most of these OCP components have been changed in the past to avoid this use of
+the `Always` pull policy. Recently the OVN pre-puller has also been changed (see
+this [bug](https://issues.redhat.com/browse/OCPBUGS-13219) for details).  To
+prevent bugs like this happening in the future and make the reboots of nodes and
+upgrades less fragile we should have a test that gates the OpenShift release and
+that verifies that reboots of nodes and upgrades can be performed without a
+registry server. One way to ensure this is to run in CI an admission hook that
+rejects/warns about any spec that uses the `Always` pull policy.
```

Comment:
It may be worth thinking about adding something to the test that blocks access to the registry till the update is done.

Reply:
Yeah that was the only way I could think of to test this. Blackhole registry.access.rdhat.com registry.redhat.io and quay.io after images are pinned to all nodes, rolling restart of all nodes or upgrade.